File: src/main/java/io/github/bot/terminal/application/domains/notificarion_api/entity/SlackNotificationApi.java
package io.github.bot.terminal.application.domains.notificarion_api.entity;


import io.github.bot.terminal.application.domains.integrations.DocumentFile;
import io.github.bot.terminal.application.domains.integrations.slack.SlackApiClient;
import io.github.bot.terminal.application.domains.integrations.slack.models.Channel;
import io.github.bot.terminal.application.domains.integrations.slack.models.SlackMessage;
import io.github.bot.terminal.application.domains.integrations.slack.requests.SendMessageRequest;
import io.github.bot.terminal.application.domains.integrations.slack.responses.GetChannelHistoryResponse;
import io.github.bot.terminal.application.domains.integrations.slack.responses.GetChannelsResponse;
import io.github.bot.terminal.application.domains.notificarion_api.repository.NotificationApiRepository;

import java.io.IOException;
import java.util.*;
import java.util.stream.Collectors;


public final class SlackNotificationApi implements NotificationApi<SlackNotificationApiDetails> {

    private final SlackNotificationApiDetails details;
    private final NotificationApiRepository repository;
    private final SlackApiClient slackApiClient;

    public SlackNotificationApi(SlackNotificationApiDetails details,
                                NotificationApiRepository repository,
                                SlackApiClient slackApiClient) {
        this.details = details;
        this.repository = repository;
        this.slackApiClient = slackApiClient;

    }

    public SlackNotificationApiDetails getDetails() {
        return details;
    }

    @Override
    public List<Message> getLastMessages() {

        Optional<Channel> chanelOptional = getChanel();
        if (!chanelOptional.isPresent()) {
            return Collections.emptyList();
        }
        Channel chanel = chanelOptional.get();

        GetChannelHistoryResponse channelHistory
                = slackApiClient.getChannelHistory(details.getToken(),
                chanel.getId(), String.format("%.0f", details.getOldest()));
        if (!Boolean.TRUE.equals(channelHistory.getOk())) {
            return Collections.emptyList();
        }

        Collection<SlackMessage> messages = channelHistory.getMessages();
        if (messages == null || messages.isEmpty()) {
            return Collections.emptyList();
        }

        Double last = messages.stream()
                .sorted(Comparator.comparingDouble(SlackMessage::getTs).reversed())
                .map(SlackMessage::getTs)
                .findFirst().orElse(null);
        if (last == null) {
            details.setOldest(new Double(new Date().toInstant().getEpochSecond()));
        } else {
            details.setOldest(last + 1D);
        }
        repository.update(details);

        return messages.stream().map(slackMessage -> new Message().setText(slackMessage.getText()))
                .collect(Collectors.toList());
    }

    @Override
    public void sendMessage(Message message) {
        getChanel().ifPresent(channel -> {
            SendMessageRequest request = new SendMessageRequest();
            request.setText(message.getText());
            request.setChannel(channel.getId());
            slackApiClient.sendMessage(details.getToken(), request);
        });
    }

    @Override
    public void sendDocument(DocumentFile file) {
        getChanel().ifPresent(channel -> {
            try {
                slackApiClient.sendDocument(details.getToken(), channel.getId(), file);
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        });
    }

    private Optional<Channel> getChanel() {
        GetChannelsResponse channelsResponse = slackApiClient.getСhannels(details.getToken());
        if (!Boolean.TRUE.equals(channelsResponse.getOk())) {
            return Optional.empty();
        }
        return channelsResponse.getChannels().stream()
                .filter(c -> details.getChanel().equals(c.getName()))
                .findFirst();
    }

    @Override
    public void delete() {
        repository.deleteById(details.id);
    }
}

File: src/main/kotlin/io/github/bot/terminal/application/domains/notificarion_api/entity/SlackNotificationApi.kt
package io.github.bot.terminal.application.domains.notificarion_api.entity

import io.github.bot.terminal.application.domains.integrations.DocumentFile
import io.github.bot.terminal.application.domains.integrations.slack.SlackApiClient
import io.github.bot.terminal.application.domains.integrations.slack.models.Channel
import io.github.bot.terminal.application.domains.integrations.slack.requests.SendMessageRequest
import io.github.bot.terminal.application.domains.notificarion_api.repository.NotificationApiRepository
import java.io.IOException

class SlackNotificationApi(override val details: SlackNotificationApiDetails,
                           private val repository: NotificationApiRepository,
                           private val slackApiClient: SlackApiClient) : NotificationApi {


    override val id: String = details.id

    override val lastMessages: List<Message>
        get() {
            chanel?.let {
                val channelHistory = slackApiClient.getChannelHistory(details.token,
                        it.id, String.format("%.0f", details.oldest))
                if (true != channelHistory.ok) {
                    return emptyList()
                }
                val messages = channelHistory.messages
                if (messages.isNullOrEmpty()) {
                    return emptyList()
                }

                messages.sortedBy { it.ts }.reversed()
                        .map { it.ts }
                        .firstOrNull()
                        ?.let {
                            details.oldest = (it + 1).toLong()
                        }
                repository.update(details)

                return messages
                        .map { Message(it.text) }
                        .toList()
            }?: return emptyList()
        }

    override fun sendMessage(message: Message) {
        chanel?.let { channel: Channel ->
            val request = SendMessageRequest()
            request.text = message.text
            request.channel = channel.id
            slackApiClient.sendMessage(details.token, request)
        }
    }

    override fun sendDocument(file: DocumentFile) {
        chanel?.let { channel: Channel ->
            try {
                slackApiClient.sendDocument(details.token, channel.id, file)
            } catch (e: IOException) {
                throw RuntimeException(e)
            }
        }
    }



    private val chanel: Channel?
        get() {
            val channelsResponse = slackApiClient.getСhannels(details.token)
            return if (true != channelsResponse.ok) {
                null
            } else channelsResponse.channels.firstOrNull{
                c: Channel -> details.chanel == c.name
            }
        }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as SlackNotificationApi

        if (details != other.details) return false
        if (repository != other.repository) return false
        if (slackApiClient != other.slackApiClient) return false
        if (id != other.id) return false

        return true
    }

    override fun hashCode(): Int {
        var result = details.hashCode()
        result = 31 * result + repository.hashCode()
        result = 31 * result + slackApiClient.hashCode()
        result = 31 * result + id.hashCode()
        return result
    }
}